lib/std/unicode.zig

```diff
@@ -196,22 +196,119 @@ pub fn utf8CountCodepoints(s: []const u8) !usize {
     return len;
 }
 
-pub fn utf8ValidateSlice(s: []const u8) bool {
+// Ported from go, which is licensed under a BSD-3 license.
+// https://golang.org/LICENSE
+//
+// https://golang.org/src/unicode/utf8/utf8.go
+/// Returns true if the input consists entirely of UTF-8 runes
+pub fn utf8ValidateSlice(input: []const u8) bool {
+    var p = input;
+
+    // Fast path. Check for and skip 8 bytes of ASCII characters per iteration.
+    while (p.len >= 8) {
+        const first32 = @as(u32, p[0]) | @as(u32, p[1]) << 8 | @as(u32, p[2]) << 16 | @as(u32, p[3]) << 24;
+        const second32 = @as(u32, p[4]) | @as(u32, p[5]) << 8 | @as(u32, p[6]) << 16 | @as(u32, p[7]) << 24;
+        if ((first32 | second32) & 0x80808080 != 0) {
+            // Found a non ASCII byte
+            break;
+        }
+        p = p[8..];
+    }
+
+    const RuneSelf = 0x80;
+    const locb = 0b10000000;
+    const hicb = 0b10111111;
+
+    // These names of these constants are chosen to give nice alignment in the
+    // table below. The first nibble is an index into acceptRanges or F for
+    // special one-byte cases. The second nibble is the Rune length or the
+    // Status for the special one-byte case.
+    const xx = 0xF1; // invalid: size 1
+    const as = 0xF0; // ASCII: size 1
+    const s1 = 0x02; // accept 0, size 2
+    const s2 = 0x13; // accept 1, size 3
+    const s3 = 0x03; // accept 0, size 3
+    const s4 = 0x23; // accept 2, size 3
+    const s5 = 0x34; // accept 3, size 4
+    const s6 = 0x04; // accept 0, size 4
+    const s7 = 0x44; // accept 4, size 4
+
+    const first = [256]u8{
```

Comment:
1. This lookup table can be generated at comptime: `const first = comptime [256]u8 ...`.
2. The first 128 values can be generated like so: `[_]u8{as} ** 128`.